lib/services/page-service.ts

```diff
@@ -0,0 +1,152 @@
+                })
+                .catch((promiseError) => {
+                    reject(promiseError);
+                });
+        });
+    }
+
+    private getAppPath(appPath: string) {
+        return new Promise((resolve, reject) => {
+            if (!util.path.isAbsolute(appPath)) {
+                reject("Path must be absolute");
+            }
+
+            appPath = util.path.normalize(appPath);
+
+            util.fs.lstat(appPath, (statErr: any, stats: any) => {
+                if (statErr) {
+                    reject(statErr);
+                }
+
+                if (!stats.isDirectory()) {
+                    reject("Not a valid app folder!");
+                } else {
+                    resolve(appPath);
+                }
+            });
+
+        });
+    }
+
+    private readAppsPackageJson(appPath: string) {
+        return new Promise((resolve, reject) => {
+            util.fs.readdir(appPath, (readErr: any, files: any) => {
+                if (readErr) {
+                    reject(readErr);
+                }
+
+                if (files.indexOf("package.json") > -1) {
+                   const packageJsonPath = util.path.join(appPath, "package.json");
+
+                   util.fs.readFile(packageJsonPath, "utf8", (fileErr: any, data: any) => {
+                       if (fileErr) {
+                           reject(readErr);
+                       }
+                       resolve(data);
+                   });
+                } else {
+                    reject("Missing package.json file in path " + appPath);
+                }
+
+            });
+        });
+
+    }
+
+    private downloadPage(pageName: string, flavor: string) {
+        const templatesDir = util.path.join(__dirname, "../templates");
+        const command = "git";
+        const commandArgs: Array<any> = ["clone"];
+
+        return new Promise((resolve, reject) => {
+            util.fs.ensureDir(templatesDir)
+                .then(() => {
+                    return util.pageExists(templatesDir, pageName);
+                })
+                .then((pageExists: boolean) => {
+                    if (!pageExists) {
+                        return util.getPageTemplatesBaseUrl(flavor);
+                    } else {
+                        // TODO If the page template indeed still exists in the tmp folder
+                        // just copy the content to the app dir
+                        resolve("Page Already exists");
+                    }
+                })
+                .then((baseUrl: string) => {
+                    baseUrl = baseUrl + ".git";
+                    commandArgs.push(baseUrl);
+                    commandArgs.push(templatesDir);
+
+                    const process = util.childProcess.spawn(command, commandArgs);
+                    process.on("close", (code) => {
+                        if (code !== 0) {
+                            reject(`child process exited with code ${code}`);
```

Comment:
Better make the rejection reason an instanceof Error.

Reply:
Done